pkg/ignore/types.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignore
+
+import (
+	"os"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Ignorer  helps identifying if a BundleFile needs to be ignored
+type Ignorer interface {
+	// Ignore checks file has to be ignored or not, returns true if it needs to be ignored
+	Ignore(path string, fi os.FileInfo) Ignoreable
+	// Patterns returns the pattern associated with interface
+	Patterns() *sets.String
+}
+
+// IgnoredType helps classifying the ignored file.
+// This helps in WalkFunc to do filepath.SkipDir for more efficient traversals
+type IgnoredType string
+
+const (
+
+	// Current holds all direct files that are ignored
+	Current IgnoredType = "Current"
+
+	// Transitive holds all the directory types that are ignored
+	// these types might have directory ignored, but can have invert patterns
+	Transitive IgnoredType = "Transitive"
+)
+
+// Ignoreable holds whether the file or directory is qualified to be ignored
+type Ignoreable struct {
+	Ignored bool
+	Type    IgnoredType
+}
```

Comment:
If we introduce an enum value for `NotIgnored` or `Include` then we should be able to eliminate the need for both the bool and the struct.

Reply:
I see we can eliminate bool but thinking how to remove the struct , as we need to pass back the type of ignore `Current` or `Transitive`

Reply:
did i miss what you  say here ?

Reply:
Was this something you had in mind with this coment

types.go

```go
// Ignoreable holds whether the file or directory is qualified to be ignored
type Ignoreable bool

// IgnoredType helps classifying the ignored file.
// This helps in WalkFunc to do filepath.SkipDir for more efficient traversals
type IgnoredType string

const (

	// Current holds all direct files that are ignored
	Current IgnoredType = "Current"

	// Transitive holds all the directory types that are ignored
	// these types might have directory ignored, but can have invert patterns
	Transitive IgnoredType = "Transitive"

	// NotIgnored flags the file should not be ignored
	NotIgnored Ignoreable = false
)
```

and `Ignore` to be 

```go
// Ignore implements Ignorer, for no dockerignore cases, where only .git is ignored
func (i *defaultIgnorer) Ignore(path string, fi os.FileInfo) (Ignoreable, IgnoredType) {
	bundleFile := &bundleFile{
		name:    fi.Name(),
		path:    path,
		rootDir: directory,
		isDir:   fi.IsDir(),
	}

	if directory != path {
		if !checkIfParentIncluded(path, fi) {
			ig := bundleFile.ignore()
			if fi.IsDir() {
				return Ignoreable(ig), Current
			} else {
				return Ignoreable(ig), Transitive
			}

		}
	}

	return NotIgnored, Transitive
}
```